examples/Custom_Hooks/campaign_stopping.py

```diff
@@ -0,0 +1,238 @@
+    NumericalDiscreteParameter(name="Temp_C", values=[90, 105, 120], tolerance=2),
+    NumericalDiscreteParameter(name="Concentration", values=[0.057, 0.1, 0.153]),
+]
+scenarios = {
+    "Average Non-Stopped": Campaign(
+        searchspace=SearchSpace.from_product(parameters=parameters), objective=objective
+    )
+}
+
+results = simulate_scenarios(
+    scenarios,
+    lookup,
+    batch_size=BATCH_SIZE,
+    n_doe_iterations=N_DOE_ITERATIONS,
+    n_mc_iterations=N_MC_ITERATIONS,
+)
+
+
+### Prepare the Hook for Stopping a Campaign
+
+
+# First, we define a custom exception to identify stopping
+class CampaignStoppedException(Exception):
+    """The campaign should be stopped."""
+
+
+# Now we define a hook that analyzes the PI.
+# PI is a traditional acquisition function that measures the probability of improvement
+# for a given point. We can utilize this to create a hook that stops the campaign by
+# throwing an exception.
+# As stopping criterion, we will count the number of candidates with a PI over 1% and
+# will terminate the campaign if there are not enough points that fulfill this.
+
+PI_THRESHOLD = 0.01  # PI of 1% to identify still promising points
+PI_REQUIRED_FRACTION = 0.1  # 10% of candidates must be over the threshold
+
+
+def stop_on_PI(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+) -> None:
+    """Raise an exception if the PI based stopping criterion is fulfilled."""
+    if searchspace.type != SearchSpaceType.DISCRETE:
+        raise TypeError(
+            f"{searchspace.type} search spaces are not supported yet. "
+            f"Currently only DISCRETE search spaces are accepted."
+        )
+    train_x = searchspace.transform(measurements)
+    train_y = objective.transform(measurements)
+    acqf = ProbabilityOfImprovement()
+
+    botorch_acqf = acqf.to_botorch(self.surrogate_model, searchspace, train_x, train_y)
+    _, candidates_comp_rep = searchspace.discrete.get_candidates(
+        allow_repeated_recommendations=self.allow_repeated_recommendations,
+        allow_recommending_already_measured=self.allow_recommending_already_measured,
+    )
+    comp_rep_tensor = to_tensor(candidates_comp_rep).unsqueeze(1)
+    acqf_values = botorch_acqf(comp_rep_tensor)
+
+    n_pis_over = torch.sum(acqf_values > PI_THRESHOLD)
+    n_pis_over_required = len(candidates_comp_rep) * PI_REQUIRED_FRACTION
+    if n_pis_over < n_pis_over_required:
+        raise CampaignStoppedException(
+            f"Less than {PI_REQUIRED_FRACTION*100:.0f}% of candidates are above the PI "
+            f"threshold of {PI_THRESHOLD*100:.0f}% - Stopping the campaign."
+        )
+
+
+# Now we attach the hook to the ``.recommend`` function of the ``BotorchRecommender``
+# we intend to use. This will attach the hook to all future instances of that class,
+# which is OK in this example as we do not plan on creating more afterwards. It is also
```

Comment:
Isn't it even *necessary* in this example due to the inner workings of the `simulate` function?

Reply:
kind of true, but perhaps this can be considered a design flaw that hopefully changes in the future either by removing deepcopies or slots or whatever. So I woudlnt write that `necessary` int stone here

Reply:
Have reorganized the content quite a bit. Now, the explanation appears cleanly separated in an admonition so that we can easily remove it in case we find a fix. Until then, I also see no better solution ...